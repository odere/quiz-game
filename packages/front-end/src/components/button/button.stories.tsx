import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import Button from './button.view';
import { ButtonSize } from './button.types'


const onClick = () => console.log('Click!');

const Template: ComponentStory<typeof Button> = (args: any) => <Button {...args} />;

export const Default = Template.bind({});

Default.args = {
    label: 'Button',
};

export const Primary = Template.bind({});

Primary.args = {
    label: 'Button',
    onClick,
    primary: true,
};

export const Secondary = Template.bind({});

Secondary.args = {
    label: 'Button',
    onClick,
    primary: false,
};

export const Small = Template.bind({});

Small.args = {
    label: 'Button',
    onClick,
    size: ButtonSize.small,
};

export const Large = Template.bind({});

Large.args = {
    label: 'Button',
    onClick,
    size: ButtonSize.large,
};

export default {
    title: 'Components/Button',
    component: Button,    
    argTypes: {
        size: {
            options: [ButtonSize.small, ButtonSize.large],
            control: { type: 'radio' }
        },
    },
} as ComponentMeta<typeof Button>
