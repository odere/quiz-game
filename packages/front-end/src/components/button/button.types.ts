export enum ButtonSize {
	small = 'small',
	large = 'large',
}

export interface ButtonStyledProps {
	size?: ButtonSize
	primary?: boolean
}

export interface ButtonProps extends ButtonStyledProps {
	label: string
	onClick?(): void
}
