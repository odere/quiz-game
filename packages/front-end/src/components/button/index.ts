import Button from './button.view'

export default Button
