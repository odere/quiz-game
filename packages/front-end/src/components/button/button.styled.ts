import styled from 'styled-components'

import { ButtonStyledProps, ButtonSize } from './button.types'

import {
	primary,
	primaryDark,
	secondary,
	secondaryDark,
} from '../../constants/colors'

export const StyledButton = styled.button`
	background: transparent;
	border: none;
	color: inherit;
	font-size: 1.5em;
	padding: 0.5em;
	min-width: 5em;
`

export const StyledButtonContainer = styled.div`
	background: ${(props: ButtonStyledProps) =>
		props.primary ? primaryDark : secondary};
	border-radius: 0.15em;
	color: white;
	display: inline-block;
	text-align: center;
	font-size: ${(props: ButtonStyledProps) =>
		props.size === ButtonSize.small ? '1em' : '3em'};
	transition: 200ms;
	pointer-events: ${(props: ButtonStyledProps) =>
		props.isDisable ? 'none' : 'inherite'};
	filter: ${(props: ButtonStyledProps) =>
		props.isDisable ? 'grayscale(75%)' : 'none'};

	&:hover {
		background: ${(props: ButtonStyledProps) =>
			props.primary ? primary : secondaryDark};
	}
`
