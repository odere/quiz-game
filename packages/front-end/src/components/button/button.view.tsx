import React from 'react'

import { StyledButton, StyledButtonContainer } from './button.styled'
import { ButtonProps, ButtonSize } from './button.types'

const Button: React.FC<ButtonProps> = props => {
	const { label, onClick, primary = true, size = ButtonSize.small } = props
	return (
		<StyledButtonContainer size={size} primary={primary}>
			<StyledButton onClick={onClick}>{label}</StyledButton>
		</StyledButtonContainer>
	)
}

export default Button
