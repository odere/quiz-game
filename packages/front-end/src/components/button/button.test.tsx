import React from "react"
import { render, fireEvent } from "@testing-library/react"
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Button from "./button.view"
import { StyledButtonContainer } from "./button.styled"

describe("components/Button:", () => {
    it("should be a button", () => {
        render(<Button label="Button" />)
        const headerElement = document.getElementsByTagName("button")[0]
        expect(headerElement).toBeInTheDocument()
    })

    it("should have a label", () => {
        const { getByText, rerender } = render(<Button label="Button" />)
        expect(getByText(/Button/i)).toBeInTheDocument()

        rerender(<Button label="OK" />)
        expect(getByText(/OK/i)).toBeInTheDocument()
    })

    it("should be Clickable", () => {
        const onClickMock = jest.fn();
        const { getByText } = render(<Button label="Button" onClick={onClickMock} />)
        
        fireEvent.click(getByText(/Button/i));
        fireEvent.click(getByText(/Button/i));

        expect(onClickMock).toBeCalledTimes(2)
    })

    it("should be able to change theme", () => {
        const treePrimary = renderer.create(<StyledButtonContainer primary />).toJSON()
        expect(treePrimary).toMatchSnapshot()

        const treeSecondary = renderer.create(<StyledButtonContainer primary={false} />).toJSON()
        expect(treeSecondary).toMatchSnapshot()
    })
})
