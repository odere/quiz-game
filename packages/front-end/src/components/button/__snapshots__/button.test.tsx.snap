// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Button: should be able to change theme 1`] = `
.c0 {
  background: #240768;
  border-radius: 0.15em;
  color: white;
  display: inline-block;
  text-align: center;
  font-size: 3em;
  -webkit-transition: 200ms;
  transition: 200ms;
  pointer-events: inherite;
  -webkit-filter: none;
  filter: none;
}

.c0:hover {
  background: #44168b;
}

<div
  className="c0"
/>
`;

exports[`components/Button: should be able to change theme 2`] = `
.c0 {
  background: #ce7642;
  border-radius: 0.15em;
  color: white;
  display: inline-block;
  text-align: center;
  font-size: 3em;
  -webkit-transition: 200ms;
  transition: 200ms;
  pointer-events: inherite;
  -webkit-filter: none;
  filter: none;
}

.c0:hover {
  background: #e3640d;
}

<div
  className="c0"
/>
`;
